instrumentation/opentelemetry-instrumentation-threading/src/opentelemetry/instrumentation/threading/__init__.py

```diff
@@ -0,0 +1,65 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=empty-docstring,no-value-for-parameter,no-member,no-name-in-module
+
+import threading  # pylint: disable=import-self
+from typing import Collection
+
+from opentelemetry import context, trace
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.threading.package import _instruments
+from opentelemetry.instrumentation.threading.version import __version__
+from opentelemetry.trace import (
+    get_current_span,
+    get_tracer,
+    get_tracer_provider,
+)
+
+
+class _InstrumentedThread(threading.Thread):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._parent_span = None
+
+    def start(self):
+        self._parent_span = get_current_span()
+        super().start()
+
+    def run(self):
```

Comment:
From [the python docs](https://docs.python.org/3/library/threading.html#thread-objects)

> There are two ways to specify the activity: by passing a callable object to the constructor, or by overriding the [run()](https://docs.python.org/3/library/threading.html#threading.Thread.run) method in a subclass.

This looks good for the first way (passing a callable) but it won't work right for second way of subclassing Thread, correct?

Reply:
Unless they call `super().run()` in the overriden `run`.

Reply:
Yes but users shouldn't have to modify their code for this instrumentation to work

Reply:
Ah, I see, they may not already have that call in their code.

Reply:
It's OK if we leave this out of scope for now, we should just document the caveat

Reply:
In that case, this may work:

``` python
import threading
from time import sleep


class InstrumentedThread(threading.Thread):

    def __getattribute__(self, name):
        if name == "run":
            print("instrumenting")

        return object.__getattribute__(self, name)


threading.Thread = InstrumentedThread

from threading import Thread  # noqa


def function():
    for _ in range(5):
        print("function")
        sleep(0.1)


user_thread_0 = Thread(target=function)
user_thread_0.start()


class UserThread(Thread):

    def run(self):
        for _ in range(5):
            print("run")
            sleep(0.1)


user_thread_1 = UserThread()
user_thread_1.start()
```
```
instrumenting
function
instrumenting
run
function
run
function
run
function
run
function
run
```

Reply:
I would suggest to leave this out of scope for the initial version and address it separately.